### DISCUSSION TITLE
Major rewrites to jura

### DISCUSSION THREAD

**Discussion by Nooberieno** [Nooberieno/jura#10] [Announcements]
# The future of jura
jura has been my first real side project and it has been coming along much faster then expected, recently due to personal things and development on other projects, development on jura has slowed down a lot. I will be returning to jura soon, with some major rewrites in order to fix some things I have done incorrectly in my opinion. I now know a lot more of how things can be done better and will use these rewrites to improve the jura experience, mostly on the backend. This announcement is to show what will be coming in those rewrites in order to expand jura. 

## 1. Compliance with XDG and POSIX
Currently, jura's config is located in the HOME directory, which just cluters up your HOME directory for no good reason. In the rewrite I will be moving the config to the .config folder to comply with XDG Base Directory Specification. POSIX compliance will also be achieved in the coming rewrite, though neither of these should impact the end user much.

## 2. Expansion of official support.
The reason for compliance with POSIX and XDG is to expand jura to more systems, this does not only mean more linux distrobutions, but also to other UNIX-like systems, such as macOS and FreeBSD, all of the eventual official release will be listed on the wiki. The plan is to expand to almost every distro and OS that will allow me to maintain and package the most up-to-date version of jura, so that users can downloaded officially maintained versions of jura via their package manager and also the [Linux User REpository](https://github.com/lure-sh/lure) project by Elara6331.

## 3. Modularization
jura source code is contained within a single C file. This was convenient when the source code was relatively small and made it so I did not have to include header files, but it has come to a point where it is no longer working. I may keep a branch open where a single file version of jura remains open, but the main branch of the code will have the code split up into more C files which will be compiled as a Shared Object file as the jura library, The library code will still be open sourced, but for convenience there will be header files if you wish to use jura's functionality for your own project. 

## 4. code readabilty
I have done my best to make jura's code well documented and understandable, but there is a lot of nested code, which works completely against this goal. During these rewrites I will try my absolute best to make the code less nested and more understandable for everyone.

# TLDR
The config file will move to a different place, the code will partially become a library to improve preformance and make jura more lightweigt and portable, more distro's and UNIX-like systems will official access to jura via their package manager and if you want to look at the code it should be easier to understand.

# Process
This part will tell all the thing that will happen during the rewrite phase.

## Merger
All currently open branches with useful changes will be merged and released in the upcoming 4.1 release. This release is what you will have to rely on when downloading jura until the rewritten version releases with 5.0.

## Rewrite phase
in this phase the code in the main branch in the single file of jura will be copied to a "legacy" branch, this will likely become the main branch for the single file version of jura. The code in the main branch will be rewritten to achieve the above goals and then get tested extensively to make fix most bugs and ensure jura is behaving as it should with all of the current features in tact.

## 5.0 release
When all of that is completed jura 5.0 will be released, first on github and later on the AUR. Then I will begin the process of getting this release into the repositories of other populair package managers that allow it.

# Thanks for reading
If you made it here, thanks for reading and I hope you enjoy jura.
